extra/checksums/multi/multi.factor

```diff
@@ -0,0 +1,31 @@
+! Copyright (C) 2018 Alexander Ilin.
+! See http://factorcode.org/license.txt for BSD license.
+USING: accessors checksums checksums.common destructors fry kernel
+sequences ;
+IN: checksums.multi
+
+TUPLE: multi-checksum list ;
+INSTANCE: multi-checksum block-checksum
+C: <multi-checksum> multi-checksum
+
+TUPLE: multi-state < disposable states results ;
+
+M: multi-checksum initialize-checksum-state
+    list>> [ initialize-checksum-state ] V{ } map-as
+    multi-state new-disposable swap >>states ;
+
+M: multi-state dispose*
+    states>> [ [ dispose ] each ] when* ;
+
+M: multi-state add-checksum-bytes
+    '[ [ _ add-checksum-bytes ] map! ] change-states ;
+
+M: multi-state get-checksum
```

Comment:
Why are you caching the results?

```states>> [ get-checksum ] map``

Reply:
Actually, I'm afraid that some of the disposed states may not handle the repeated `get-checksum` well. Not sure if that's a real issue, or an imagined one.

Reply:
Why would you make repeated calls to ``get-checksum``?  Seems like that's not necessarily supported by any checksum so why add caching here?

Reply:
Well, it's kind of "there" in the API: you can call `get-checksum` whenever you want. I'm not calling it twice in any of my code, but it's not explicitly forbidden in the documentation, so I support the use case. Why not call `get-checksum` several times after the data is piped? You have the finished state, you can request the result as many times as you like. The data are there, in the object.

I never saw it unsupported or forbidden or "not guaranteed", so I support it in my hashes. E.g. the `checksums.process` vocab keeps the data after the underlying process is terminated. I think our API/protocol is not very clear in this regard, so I'm trying to be on the safe side.

Reply:
That makes sense, but I'd defer the caching to the checksums you're using rather than putting it in the multi instance personally.

But if you do that, I'm not sure I would dispose the ``states``, since that's being done in ``M\ multi-state dispose``.

```factor
M: multi-state get-checksum
    dup results>> [
        dup states>> [ get-checksum ] map [ >>results ] keep
    ] unless* nip ;
```

Reply:
I think the last change you suggested is acceptable. I wanted to dispose of things as early as possible (in case user forgets to call `dispose` on the `multi-state`, and only ever calls the `get-checksum`), but I guess it mixes up the responsibilities between the two methods. I wanted the stronger, more encapsulated guarantees of the protocol, and you suggest we let the weaker implementations seep through to the surface. I can roll with that, even though it's not how I normally prefer to build systems.